execute/plugin.go

```diff
@@ -209,50 +210,55 @@ func getPendingReportsForExecution(
 				"skippedCommitRoots", skippedCommitRoots,
 			)
 		}
-		reports = filtered
-
-		lggr.Debugw("grouped reports", "selector", selector, "reports", reports, "count", len(reports))
-		sort.Slice(reports, func(i, j int) bool {
-			return reports[i].SequenceNumberRange.Start() < reports[j].SequenceNumberRange.Start()
+		filteredReports[selector] = filtered
+
+		lggr.Debugw("grouped reports",
+			"selector", selector,
+			"reports", filteredReports[selector],
+			"count", len(filteredReports[selector]))
+		sort.Slice(filteredReports[selector], func(i, j int) bool {
+			return filteredReports[selector][i].SequenceNumberRange.Start() <
+				filteredReports[selector][j].SequenceNumberRange.Start()
 		})
 		// todo: remove this logs after investigating whether the sorting above can be safely removed
-		lggr.Debugw("sorted reports", "selector", selector, "reports", reports, "count", len(reports))
+		lggr.Debugw("sorted reports",
+			"selector", selector,
+			"reports", filteredReports[selector],
+			"count", len(filteredReports[selector]))
 
-		ranges, err := computeRanges(reports)
+		ranges, err := computeRanges(filteredReports[selector])
 		if err != nil {
-			return nil, nil, nil, fmt.Errorf("compute report ranges: %w", err)
-		}
-
-		// Get both finalized and unfinalized executed messages
-		finalizedMsgSet := mapset.NewSet[cciptypes.SeqNum]()
-		allExecutedMsgSet := mapset.NewSet[cciptypes.SeqNum]()
-
-		for _, seqRange := range ranges {
-			// Get all executed messages
-			allMessages, err := ccipReader.ExecutedMessages(ctx, selector, seqRange, primitives.Unconfirmed)
-			if err != nil {
-				return nil, nil, nil, fmt.Errorf("get %d executed messages in range %v: %w", selector, seqRange, err)
-			}
-			allExecutedMsgSet = allExecutedMsgSet.Union(mapset.NewSet(allMessages...))
-
-			// Get finalized messages
-			finalizedMessages, err := ccipReader.ExecutedMessages(ctx, selector, seqRange, primitives.Finalized)
-			if err != nil {
-				return nil, nil, nil, fmt.Errorf("get finalized %d executed messages in range %v: %w", selector, seqRange, err)
-			}
-			finalizedMsgSet = finalizedMsgSet.Union(mapset.NewSet(finalizedMessages...))
+			return nil, nil, nil,
+				fmt.Errorf("compute report ranges: %w", err)
 		}
+		rangesBySelector[selector] = ranges
+	}
+	// Get all executed messages
+	allMessages, err := ccipReader.ExecutedMessages(ctx, rangesBySelector, primitives.Unconfirmed)
+	if err != nil {
+		return nil, nil, nil,
+			fmt.Errorf("get executed messages in range %v: %w", rangesBySelector, err)
+	}
+	// Get finalized messages
+	finalizedMessages, err := ccipReader.ExecutedMessages(ctx, rangesBySelector, primitives.Finalized)
+	if err != nil {
+		return nil, nil, nil,
+			fmt.Errorf("get finalized executed messages in range %v: %w", rangesBySelector, err)
+	}
 
+	for selector, reports := range filteredReports {
```

Comment:
This function is actually really hard to follow despite it doing a supposedly simple thing. Here's my attempt at a refactoring into pure funcs that are separately testable and makes it easier to digest: https://github.com/smartcontractkit/chainlink-ccip/pull/759